### ISSUE TITLE
Suggestion to add initialValues prop for Server Side Rendering

### ISSUE THREAD

**Issue by markau** [sslotsky/violet-paginator#7]
[redux-form](https://github.com/erikras/redux-form) takes an `initialValues` prop, which hydrates the state from the server so that the page appears complete on first render. Would this be hard to achieve with [violet-paginator](https://github.com/sslotsky/violet-paginator)?

**Comment by sslotsky** [comment]
I don't think it would be hard to add support for that. We could allow the `initialize` action to accept some optional initial values. The pagination wrapper could also accept this optional prop and pass it along to `initialize`. 

Feel free to submit a PR if I can't get to it first.

**Comment by sslotsky** [comment]
@markau I submitted #8 as a proposed solution for this. I haven't tested it yet whatsoever. Do you want to see if it works for you? The idea is you supply a `preloaded` prop to the paginated component, and that would be an object like this:

```javascript
const preloaded = {
  results,
  totalCount,
  page
};
```

That should allow the component to render with that data and also prevent it from performing a fetch when it initially renders.

**Comment by markau** [comment]
Yes, nice work, that is doing the job. It does cause a re-render of the page though - i.e. the grid header appears instantly on page load, and then a short time later the data appears. It's faster than waiting for the API with the spinner. I am passing the `preloaded` data via `mapStateToProps` to the connected component FWIW.

**Comment by markau** [comment]
Oh and it doesn't paginate on that initial SSR load (it displays all results). Does `preloaded` need a 'number of items per page' property?

**Comment by sslotsky** [comment]
My thinking was that you would just pass in a single page of data. `violet-paginator` is designed to request new pages of data from the `fetch` function that you provide. It's probably not a good idea to pass in your entire result set, because if your result set gets large your query will take too long to execute. So something more like this would probably fit the bill in terms of the data that your server is returning:

```javascript
const query = SomeModelClass.all
const totalCount = query.count()
const page, perPage = getPageParameters()
const results = query.paginate({ page, perPage })

return json({
  results,
  totalCount,
  page
})
```

**Comment by markau** [comment]
Ok that seems obvious now that you mention it. Thanks.

It is working well. 

This is beyond my React/JavaScript knowledge, so if you happened to know why the pagination results (i.e. the numbered buttons), and the `preloaded` data, force a re-render - both follow the initial server page render by about half a second - then I'd be keen to explore it further.

**Comment by sslotsky** [comment]
Sure, I will have to do some testing and see what I can see. I suspect the result would be the same without server rendering, i.e. just using the `preloaded` prop from a parent component. So I can try to recreate it that way, or you can link me to your repo if you like.

**Comment by sslotsky** [comment]
> I have seen this issue where components must call the window namespace (which doesn't exist on the server, only after the client JavaScript runs.)

I confess that I am inexperienced with server rendering. If you think the above point is relevant, can you elaborate on it or provide some resources?

**Comment by sslotsky** [comment]
@markau I've been hacking on a server rendering example today and I've confirmed my suspicions as to the cause of the re-render. The first problem is that the preloaded data doesn't get inserted directly into the components, but instead it gets passed into an action creator that triggers a state update. After the state update, the components re-render with the updated state that contains your preloaded data.

I was able to fix the datatable by inserting those props directly into the component if they are present and we are uninitialized. However, this doesn't currently work for the pagination controls because the props that they rely on are getting updated as a result of the `initialize` action.

In other words, it may take me a little bit of refactoring to solve this problem, but I think I'm pretty close to a solution. Thanks for your patience.

**Comment by markau** [comment]
Sorry for the poor response time, and don't stress because this library is so good, with or without SSR.

Now that you outlined the state update that makes sense as to why it would re-render. I have learnt enough about React and Redux to understand that; but I probably wouldn't have been able to work it out. So, thanks.

Take your time - I am happy to do any testing, just let me know.

**Comment by sslotsky** [comment]
@markau I think I got it. My last commit includes an example server rendering app that seems to work correctly with no re-render. Please let me know if it works for you. I'll have to add some automated testing before merging, but until then you could reference the commit hash in your package.json

```javascript
"violet-paginator": "github:sslotsky/violet-paginator#ef40fccc11a2310258f2fa296171d73cbc94659d"
```

**Comment by markau** [comment]
No worries! I'll explore this over the next day or so. Just to be a problem user - I am using [ReactJS.NET,](https://reactjs.net/) not Node/Express, which presents some limitations - for example, I need to pass the preloaded data in as a prop and not via the `window.__PRELOADED_STATE__` method. This is not ideal, but works for my scenario. I have had to do this workaround with some other libraries; it shouldn't require any changes to your commit so far.

On first glance, the main thing I need to work out is if my `preloaded` data coming in (via prop) is shaped liked this:

````
const initData = {
      "results": [
        {
          "id": 0,
          "hop": "chinook",
          "AAU": 5,
        },
        {
          "id": 1,
          "hop": "cascade",
          "AAU": 5,
        }
      ],
  totalCount: 2
};
````

I see that I need to instantiate the `Violet-Paginator` reducer with this:

````
pagination: List(preloadedState.pagination)
````

instead of what I have currently:

````
pagination: pagination
````

Therefore, my `preloadedState` object needs to be more than just the `results `object.

**Comment by sslotsky** [comment]
Maybe I'm reading you wrong but it sounds like your conflating `preloadedState` with the `preloaded` data. The `preloadedState` object should be what's returned by `store.getState()`. 
You are correct that `preloadedState.pagination` will have to be wrapped in a `List` when you create your store on the client side. This is a little janky to be sure, but there's no good way around it at this time since I didn't anticipate the difficulties with using immutable state together with server rendering. `violet-paginator` expects `store.getState().pagination` to contain the reducer, so we're kind of boxed into this setup for now.

**Comment by markau** [comment]
[Ahh yes, thank you for the clarification](https://media.giphy.com/media/VXCPgZwEP7f1e/giphy.gif).

It seems with the benefit of Node/React/Redux on the server, you are passing in the reducer as an object (i.e. `preloadedState.pagination`). That is pretty sweet. I will have to build `preloadedState.pagination` based on the `preloaded` data.

I feel I have taken enough of your time, and I might just need to spend some time getting stuck into it.

Thank you for your efforts so far!

**Comment by sslotsky** [comment]
Very well, I'll leave you with a final tip and wish you good luck.

`preloadedState.pagination` is _not_ built based on the `preloaded` data, as you say. In my case, I assembled the `preloaded` object out of some dummy data here:

https://github.com/sslotsky/violet-paginator/pull/8/files#diff-ec1f2d30c0816883347c4f8dfbe79f8cR26

The `preloadedState`, on the other hand, comes from `store.getState()` after using `combineReducers` and `createStore` to create the redux store. It then gets passed into the server template.

https://github.com/sslotsky/violet-paginator/pull/8/files#diff-07409a3d26b27197506269128d994d73R31

I basically just followed the [redux server rendering example)(http://redux.js.org/docs/recipes/ServerRendering.html) to create the store and save it to `window.__PRELOADED_STATE__`. Then I assembled my `preloaded` object and passed it directly into the `VioletPaginator` and `VioletDataTable` components.

Hopefully that clears some things up further. Best of luck!

**Comment by markau** [comment]
Hey Sam, I agree that the issue is resolved. Thanks for implementing it. I'm sure I will get it working when I get some time; just a matter of priorities.

**Comment by sslotsky** [comment]
Hi @markau , I know this issue's closed, but I just finished working up a more realistic example of SSR with `violet-paginator` for a personal project that I'm hacking on. As I worked through it I realized that passing some `preloaded` data into the component itself isn't really feasible in the typical setup. However, if you do something like this on the server:

```javascript
      const pageActions = composables({ listId: 'posts', preloaded: payload })
      return store.dispatch(pageActions.initialize())
```

then the data will be correctly preloaded on the client. Here's a full SSR example using `react-router`

```javascript
import React from 'react'
import DevTools from 'CLIENT/DevTools'
import { match, RouterContext, createMemoryHistory } from 'react-router'
import { Provider } from 'react-redux'
import { renderToString } from 'react-dom/server'
import { syncHistoryWithStore, routerMiddleware } from 'react-router-redux'
import thunk from 'redux-thunk'
import { compose, createStore, applyMiddleware } from 'redux'

import routes from 'CLIENT/config/routes'
import reducer from 'CLIENT/reducer'
import template from './template'
import Browse from 'MODULES/posts/components/Browse'
import * as posts from 'SERVICES/posts'
import { composables } from 'violet-paginator'

const preload = (store, query, params) => ({
  [Browse]: () => {
    return posts.list().then(payload => {
      const pageActions = composables({ listId: 'posts', preloaded: payload })
      return store.dispatch(pageActions.initialize())
    })
  }
})

export default function(req, resp) {
  const memoryHistory = createMemoryHistory(req.path)
  const router = routerMiddleware(memoryHistory)

  const store = createStore(
    reducer,
    compose(applyMiddleware(router, thunk), DevTools.instrument())
  )

  const history = syncHistoryWithStore(memoryHistory, store)

  match({ history, routes, location: req.url }, (error, redirectLocation, renderProps) => {
    if (error) {
      res.status(500).send(error.message)
    } else if (redirectLocation) {
      res.redirect(302, redirectLocation.pathname + redirectLocation.search)
    } else if (renderProps) {
      const { components, query, params } = renderProps
      const component = components[components.length - 1]
      const handler = preload(store, query, params)[component]
      const promise = handler ? handler() : Promise.resolve()

      promise.then(() => {
        const html = renderToString(
          <Provider store={store}>
            <RouterContext {...renderProps}/>
          </Provider>
        )

        const preloadedState = store.getState()
        resp.send(template(html, preloadedState))
      })
    }
  })
}
```

**Comment by markau** [comment]
Hey @sslotsky, thanks for thinking of me and posting this. I had to drop SSR, as it was completely reliant on ReactJS.NET which is not particularly well supported (the current version causes a memory leak in the web server with no fix in sight).  Client-side is making life easier, so all good. 

Keep up the good work on `violet-paginator`. (So many downloads yet so few stars!)
